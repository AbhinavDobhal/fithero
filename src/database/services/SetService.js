/* @flow */

import type { SetSchemaType } from '../types';
import {
  ADD_SET,
  getSet,
  removeSet,
  UPDATE_SET,
} from '../../redux/modules/workouts';
import realm from '../index';
import {
  extractWorkoutKeyFromDatabase,
  getExerciseSchemaIdFromSet,
} from '../utils';
import type { DispatchType } from '../../types';

export const getMaxSetByType = (type: string) =>
  realm
    .objects('Set')
    .filtered('type = $0', type)
    .sorted([['weight', true], 'date']);

export const addSet = (
  dispatch: (DispatchType<SetSchemaType>) => void,
  set: SetSchemaType
) => {
  // Optimistic update to Redux
  dispatch(getSet(ADD_SET, set));

  realm.write(() => {
    const exerciseId = getExerciseSchemaIdFromSet(set.id);
    const exercise = realm.objectForPrimaryKey('Exercise', exerciseId);
    exercise.sets.push(set);
  });
};

export const updateSet = (
  dispatch: (DispatchType<SetSchemaType>) => void,
  updatedSet: SetSchemaType
) => {
  // Optimistic update to Redux
  dispatch(getSet(UPDATE_SET, updatedSet));

  realm.write(() => {
    const set = realm.objectForPrimaryKey('Set', updatedSet.id);
    set.weight = updatedSet.weight;
    set.reps = updatedSet.reps;
  });
};

export const deleteSet = (
  dispatch: (DispatchType<string>) => void,
  setId: string
) => {
  // Optimistic update to Redux
  dispatch(removeSet(setId));

  // Database, if last set, delete exercise, if last exercise, delete workout
  realm.write(() => {
    const setToDelete = realm.objectForPrimaryKey('Set', setId);
    realm.delete(setToDelete);
    // After deleting set, check if we need to delete the whole exercise
    const exerciseId = getExerciseSchemaIdFromSet(setId);
    const exercise = realm.objectForPrimaryKey('Exercise', exerciseId);
    if (exercise.sets.length === 0) {
      realm.delete(exercise);
      // Now we check if workout needs to be deleted too
      const workoutId = extractWorkoutKeyFromDatabase(setId);
      const workout = realm.objectForPrimaryKey('Workout', workoutId);
      if (workout.exercises.length === 0) {
        realm.delete(workout);
      }
    }
  });
};
