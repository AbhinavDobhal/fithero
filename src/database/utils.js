/* @flow */

import leftPad from 'left-pad';

export const getExerciseSchemaId = (day: string, exerciseKey: string) =>
  `${day}_${exerciseKey}`;

export const getExerciseSchemaIdFromSet = (setId: string) => {
  const parts = setId.split('_');
  return `${parts[0]}_${parts[1]}`;
};

export const getSetSchemaId = (
  day: string,
  exerciseKey: string,
  index: number
) => `${getExerciseSchemaId(day, exerciseKey)}_${leftPad(index, 3, '0')}`;

export const extractWorkoutKeyFromDatabase = (id: string) => id.split('_')[0];

export const extractExerciseKeyFromDatabase = (id: string) => id.split('_')[1];

export const extractSetIndexFromDatabase = (id: string) =>
  parseInt(id.split('_')[2], 10);
