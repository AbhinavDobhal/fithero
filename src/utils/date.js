/* @flow */

// import { NativeModules } from 'react-native';
import moment from 'moment';
import * as RNLocalize from 'react-native-localize';

// To import locales we want translations from
// import 'moment/locale/pl';
// import 'moment/locale/es';
// moment.locale(NativeModules.RNI18n.languages[0]);

// To change the starting week day to Monday
// Although locale do it for you, it can override the locale
// moment.updateLocale('en', {
//   week: {
//     dow: 1,
//   },
// });

import i18n from './i18n';
import regionDayMap from './regionDayMap';
import type { FirstDayOfTheWeekType } from '../redux/modules/settings';

// TODO this function won't work properly if we change the moment.locale
const getFirstDayOfTheWeek = (
  date: Date,
  firstDayOfTheWeek: FirstDayOfTheWeekType
) => {
  const now = moment(date);
  const firstDay = firstDayOfTheWeekToNumber(firstDayOfTheWeek);
  const firstWeekDay = now
    .clone()
    .weekday(firstDay)
    .startOf('day');
  if (!firstWeekDay.isAfter(now)) {
    return firstWeekDay;
  }
  return now
    .clone()
    .weekday(firstDay - 7)
    .startOf('day');
};

export const getCurrentWeek = (
  date: Date,
  firstDayOfTheWeek: FirstDayOfTheWeekType
) => {
  const start = getFirstDayOfTheWeek(date, firstDayOfTheWeek);
  const days = [start.clone().toDate()];

  for (let i = 1; i <= 6; i++) {
    days.push(
      start
        .clone()
        .add(i, 'day')
        .toDate()
    );
  }

  return days;
};

export const getFirstAndLastWeekday = (
  date: Date,
  firstDayOfTheWeek: FirstDayOfTheWeekType
) => {
  const now = moment(date);
  const firstDay = firstDayOfTheWeekToNumber(firstDayOfTheWeek);
  const start = getFirstDayOfTheWeek(date, firstDayOfTheWeek);
  const end = now
    .clone()
    .weekday(firstDay + 6)
    .startOf('day');

  return [start.toDate(), end.toDate()];
};

export const getShortDayInfo = (dateString: string) => {
  const momentDate = moment(dateString);
  return {
    date: momentDate.get('date'),
    day: momentDate.format('ddd'),
  };
};

export const getToday = () => moment().startOf('day');

export const isSameDay = (date: Date, today: string) =>
  moment(date).isSame(moment(today), 'day');

export const dateToWorkoutId = (date: Date | string) =>
  moment(date).format('YYYYMMDD');

export const dateToString = (date: Date) => moment(date).toISOString();

export const toDate = (dateString: string) => moment(dateString).toDate();

export const getDatePrettyFormat = (
  dateString: Date | string,
  today: string,
  short?: boolean = false
) => {
  const date = moment(dateString);
  const isToday = date.isSame(moment(today), 'day');

  if (isToday) {
    return `${i18n.t('today')}${!short ? `, ${date.format('MMMM D')}` : ''}`;
  }
  return date.format(`${!short ? 'dddd' : 'ddd'}, MMMM D`);
};

export const getDay = (day: string) =>
  moment(day)
    .startOf('day')
    .toISOString();

export const formatDate = (date: Date | string, format: string) =>
  moment(date).format(format);

export const getFirstAndLastMonthDay = (date: Date) => {
  const now = moment(date);
  const start = now.startOf('month');
  const end = now.clone().endOf('month');

  return [start.toDate(), end.toDate()];
};

export const getWeekStartByLocale = (): FirstDayOfTheWeekType => {
  const country = RNLocalize.getCountry();

  const day = regionDayMap[country];
  if (day === 0) {
    return 'sunday';
  } else if (day === 6) {
    return 'saturday';
  }
  return 'monday';
};

export const firstDayOfTheWeekToNumber = (
  firstDayOfTheWeek: FirstDayOfTheWeekType
) => {
  let firstDay = 1; // Monday
  if (firstDayOfTheWeek === 'sunday') {
    firstDay = 0;
  } else if (firstDayOfTheWeek === 'saturday') {
    firstDay = 6;
  }
  return firstDay;
};

export const getSafeTimezoneTime = (date: Date) => {
  return moment(date)
    .subtract(1, 'day')
    .toDate();
};
