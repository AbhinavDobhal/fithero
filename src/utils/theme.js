/* @flow */

import { Colors, DarkTheme } from 'react-native-paper';

export default {
  ...DarkTheme,
  colors: {
    ...DarkTheme.colors,
    primary: Colors.red500,
    accent: Colors.red500,
    icon: Colors.white,
    secondaryText: 'rgba(255, 255, 255, .7)',
    selected: Colors.teal700,
    confirm: Colors.teal700,
    delete: Colors.red700,
  },
};
