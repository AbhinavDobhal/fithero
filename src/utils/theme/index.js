/* @flow */

export { default } from './blueTheme';
