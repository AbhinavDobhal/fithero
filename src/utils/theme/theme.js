/* @flow */

import { Colors, DarkTheme } from 'react-native-paper';
import type { ThemeColorsType } from './types';
import type { ThemeType } from './withTheme';

const primary = Colors.amberA200;
const background = '#333333';

const ThemeColors: ThemeColorsType = {
  accent: primary,
  background,
  calendarSelectedDayTextColor: background,
  calendarSelectedDotColor: background,
  chip: '#424242',
  chipSelected: 'rgba(255, 255, 255, .20)',
  secondaryText: 'rgba(255, 255, 255, .7)',
  selected: 'rgba(255, 255, 255, .20)',
  surface: Colors.grey800,
  primary,
  textSelection: Colors.red300,
  toolbar: Colors.grey900,
  toolbarTint: Colors.white,
  trophy: primary,
};

const Theme: ThemeType = {
  ...DarkTheme,
  colors: {
    ...DarkTheme.colors,
    ...ThemeColors,
  },
};

export default Theme;
