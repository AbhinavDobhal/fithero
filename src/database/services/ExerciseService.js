/* @flow */

import realm from '../index';
import type { ExerciseSchemaType } from '../types';
import type { DispatchType } from '../../types';
import { extractWorkoutKeyFromDatabase } from '../utils';
import { toDate } from '../../utils/date';
import { getExercise } from '../../redux/modules/workouts';

export const addExercise = (
  dispatch: (fn: DispatchType<ExerciseSchemaType>) => void,
  exercise: ExerciseSchemaType
) => {
  // Optimistic update to Redux
  dispatch(getExercise(exercise));

  realm.write(() => {
    const workoutId = extractWorkoutKeyFromDatabase(exercise.id);
    let workout = realm.objectForPrimaryKey('Workout', workoutId);
    if (!workout) {
      workout = realm.create('Workout', {
        id: workoutId,
        date: toDate(workoutId),
      });
    }
    workout.exercises.push(exercise);
  });
};
