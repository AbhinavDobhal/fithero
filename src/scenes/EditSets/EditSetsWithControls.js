/* @flow */

import * as React from 'react';
import { FlatList, Keyboard, StyleSheet, View } from 'react-native';
import { Card } from 'react-native-paper';
import { AndroidBackHandler } from 'react-navigation-backhandler';

import type {
  WorkoutExerciseSchemaType,
  WorkoutSetSchemaType,
} from '../../database/types';
import EditSetsInputControls from './EditSetsInputControls';
import i18n from '../../utils/i18n';
import EditSetItem from './EditSetItem';
import EditSetActionButtons from './EditSetActionButtons';
import {
  extractSetIndexFromDatabase,
  getExerciseSchemaId,
  getSetSchemaId,
} from '../../database/utils';
import {
  addSet,
  deleteSet,
  getLastSetByType,
  getMaxSetByType,
  updateSet,
} from '../../database/services/WorkoutSetService';
import { addExercise } from '../../database/services/WorkoutExerciseService';
import { toDate } from '../../utils/date';
import DataProvider from '../../components/DataProvider';
import type { RealmResults } from '../../types';

type Props = {
  day: string,
  exerciseKey: string,
  exercise?: ?WorkoutExerciseSchemaType,
};

type State = {
  weight: number,
  reps: number,
  selectedId: string,
};

type ActionIncDec = (property: string, value: number) => void;

export class EditSetsWithControls extends React.Component<Props, State> {
  smallestWeightDec: ActionIncDec;
  weightDec: ActionIncDec;
  weightInc: ActionIncDec;
  smallestWeightInc: ActionIncDec;
  biggestRepsDec: ActionIncDec;
  repsDec: ActionIncDec;
  repsInc: ActionIncDec;
  biggestRepsInc: ActionIncDec;

  constructor(props: Props) {
    super(props);
    this.smallestWeightDec = this.handleIncDec.bind(this, 'weight', -0.5);
    this.weightDec = this.handleIncDec.bind(this, 'weight', -1);
    this.weightInc = this.handleIncDec.bind(this, 'weight', +1);
    this.smallestWeightInc = this.handleIncDec.bind(this, 'weight', +0.5);

    this.biggestRepsDec = this.handleIncDec.bind(this, 'reps', -2);
    this.repsDec = this.handleIncDec.bind(this, 'reps', -1);
    this.repsInc = this.handleIncDec.bind(this, 'reps', +1);
    this.biggestRepsInc = this.handleIncDec.bind(this, 'reps', +2);

    let lastSet = null;
    if (props.exercise) {
      lastSet = props.exercise.sets[props.exercise.sets.length - 1];
    } else {
      const sets = getLastSetByType(
        this.props.exerciseKey || this.props.exercise?.type
      );
      if (sets.length > 0) {
        lastSet = sets[0];
      }
    }

    this.state = {
      weight: lastSet ? lastSet.weight : 20,
      reps: lastSet ? lastSet.reps : 8,
      selectedId: '',
    };
  }

  onBackButtonPressAndroid = () => {
    if (this.state.selectedId) {
      this.setState({ selectedId: '' });
      return true;
    }
    return false;
  };

  handleIncDec = (property: string, value: number) => {
    const currentValue = this.state[property] >= 0 ? this.state[property] : 0;
    const newValue = currentValue + value;
    this.setState({ [property]: newValue > 0 ? newValue : 0 });
  };

  _onChangeWeightInput = (value: string) => {
    this.setState({ weight: value ? parseFloat(value) : -1 });
  };

  _onChangeRepsInput = (value: string) => {
    this.setState({ reps: value ? parseInt(value, 10) : -1 });
  };

  _keyExtractor = item => item.id;

  _onPressItem = (setId: string) => {
    if (this.state.selectedId === setId) {
      this.setState({ selectedId: '' });
      return;
    }
    if (this.props.exercise) {
      const { sets } = this.props.exercise;
      const set = sets.find(s => s.id === setId);
      if (set) {
        this.setState({
          selectedId: setId,
          weight: set.weight,
          reps: set.reps,
        });
      }
    }
  };

  _onAddSet = () => {
    const { day, exerciseKey, exercise } = this.props;
    const { reps, selectedId, weight } = this.state;

    let newExercise = null;

    Keyboard.dismiss();

    if (!exercise) {
      const exerciseIdDb = getExerciseSchemaId(day, exerciseKey);
      newExercise = {
        id: exerciseIdDb,
        sets: [
          {
            id: getSetSchemaId(day, exerciseKey, 1),
            weight,
            reps,
            date: toDate(day),
            type: exerciseKey,
          },
        ],
        comments: '',
        date: toDate(day),
        type: exerciseKey,
      };
      addExercise(newExercise);
    } else if (!selectedId) {
      const lastId = exercise.sets[exercise.sets.length - 1].id;
      const lastIndex = extractSetIndexFromDatabase(lastId);

      addSet({
        id: getSetSchemaId(day, exerciseKey, lastIndex + 1),
        weight,
        reps,
        date: toDate(day),
        type: exerciseKey,
      });
    } else if (selectedId) {
      updateSet({
        id: selectedId,
        weight,
        reps,
        date: toDate(day),
        type: exerciseKey,
      });
    }

    if (selectedId) {
      this.setState({ selectedId: '' });
    }
  };

  _onDeleteSet = () => {
    const { selectedId } = this.state;

    Keyboard.dismiss();

    deleteSet(selectedId);
    this.setState({ selectedId: '' });
  };

  _renderItem = ({ item, index }, maxSetId) => (
    <EditSetItem
      set={item}
      index={index + 1}
      isSelected={this.state.selectedId === item.id}
      isMaxSet={maxSetId === item.id}
      onPressItem={this._onPressItem}
    />
  );

  render() {
    const { exercise } = this.props;
    const { reps, selectedId, weight } = this.state;

    return (
      <AndroidBackHandler onBackPress={this.onBackButtonPressAndroid}>
        <View style={styles.container}>
          <Card style={styles.card}>
            <View style={styles.cardContent}>
              <EditSetsInputControls
                input={weight}
                label={i18n.t('weight_label', {
                  w: i18n.t('kg.unit', { count: 10 }),
                })}
                onChangeText={this._onChangeWeightInput}
                controls={[
                  { label: '-0.5', action: this.smallestWeightDec },
                  { label: '-1.0', action: this.weightDec },
                  { label: '+1.0', action: this.weightInc },
                  { label: '+0.5', action: this.smallestWeightInc },
                ]}
              />
              <EditSetsInputControls
                input={reps}
                label={i18n.t('reps.title')}
                onChangeText={this._onChangeRepsInput}
                controls={[
                  { label: '-2', action: this.biggestRepsDec },
                  { label: '-1', action: this.repsDec },
                  { label: '+1', action: this.repsInc },
                  { label: '+2', action: this.biggestRepsInc },
                ]}
              />
            </View>
            <EditSetActionButtons
              isUpdate={!!selectedId}
              onAddSet={this._onAddSet}
              onDeleteSet={this._onDeleteSet}
            />
          </Card>
          <DataProvider
            query={getMaxSetByType}
            args={[this.props.exerciseKey]}
            parse={(sets: RealmResults<WorkoutSetSchemaType>) =>
              sets.length > 0 ? sets[0].id : null
            }
            render={(maxSetId: string) => (
              <FlatList
                contentContainerStyle={styles.list}
                // It's possible that we delete the whole exercise so this access to .sets would be invalid
                data={exercise && exercise.isValid() ? exercise.sets : []}
                keyExtractor={this._keyExtractor}
                renderItem={propsData => this._renderItem(propsData, maxSetId)}
                extraData={[this.state.selectedId, exercise]}
                keyboardShouldPersistTaps="always"
              />
            )}
          />
        </View>
      </AndroidBackHandler>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  card: {
    marginTop: 8,
    marginHorizontal: 16,
  },
  cardContent: {
    paddingTop: 8,
    paddingHorizontal: 16,
  },
  list: {
    paddingVertical: 12,
  },
});

export default EditSetsWithControls;
