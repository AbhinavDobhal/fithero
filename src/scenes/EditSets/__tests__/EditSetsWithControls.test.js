/* @flow */

import React from 'react';
import { Keyboard } from 'react-native';
import { shallow } from 'enzyme';

import { EditSetsWithControls } from '../EditSetsWithControls';
import theme from '../../../utils/theme';
import { toDate } from '../../../utils/date';
import { deleteSet } from '../../../database/services/SetService';

jest.mock('Keyboard');

jest.mock('react-navigation-backhandler', () => ({
  AndroidBackHandler: ({ children }) => children,
}));

jest.mock('../../../database/services/SetService', () => ({
  addSet: jest.fn(),
  deleteSet: jest.fn(),
  updateSet: jest.fn(),
}));

jest.mock('../../../database/services/ExerciseService');

const date = toDate('2018-05-01T00:00:00.000Z');

describe('EditSetsWithControls', () => {
  const day = '2018-05-01T00:00:00.000Z';
  const dispatch = jest.fn();
  const exerciseKey = 'bench-press';
  const exercise = {
    id: '2018-05-01T00:00:00.000Z_bench-press',
    comments: '',
    sets: [
      {
        id: '2018-05-01T00:00:00.000Z_bench-press_001',
        reps: 8,
        weight: 100,
        date,
        type: 'bench-press',
      },
      {
        id: '2018-05-01T00:00:00.000Z_bench-press_002',
        reps: 6,
        weight: 90,
        date,
        type: 'bench-press',
      },
    ],
    date,
    type: 'bench-press',
    sort: 1,
  };

  const defaultWeight = 20;
  const defaultReps = 8;

  describe('EditSetsInputControls', () => {
    // eslint-disable-next-line flowtype/no-weak-types
    const getInputControls = (wrapper: Object, index: number) => {
      const inputControls = wrapper
        .find('withTheme(EditSetsInputControls)')
        .at(index)
        .dive();

      const Component = inputControls.props().children();
      return shallow(React.cloneElement(Component, { theme }));
    };

    it('has correct default values if there is no exercise', () => {
      const wrapper = shallow(
        <EditSetsWithControls
          day={day}
          dispatch={dispatch}
          exerciseKey={exerciseKey}
          exercisesCount={1}
          maxSetId=""
        />
      );

      expect(wrapper.state()).toEqual({
        weight: defaultWeight,
        reps: defaultReps,
        selectedId: '',
      });

      const weightControls = getInputControls(wrapper, 0);

      expect(weightControls.find('TextInput').props().value).toEqual('20');
      expect(
        weightControls
          .find('Caption')
          .children()
          .text()
      ).toEqual('Weight (kgs)');

      const repsControls = getInputControls(wrapper, 1);

      expect(repsControls.find('TextInput').props().value).toEqual('8');
      expect(
        repsControls
          .find('Caption')
          .children()
          .text()
      ).toEqual('Reps');
    });

    it('has the values of last set if we pass an exercise', () => {
      const wrapper = shallow(
        <EditSetsWithControls
          day={day}
          dispatch={dispatch}
          exerciseKey={exerciseKey}
          exercise={exercise}
          exercisesCount={1}
          maxSetId=""
        />
      );
      expect(wrapper.state()).toEqual({
        weight: exercise.sets[1].weight,
        reps: exercise.sets[1].reps,
        selectedId: '',
      });

      const weightControls = getInputControls(wrapper, 0);
      const repsControls = getInputControls(wrapper, 1);

      expect(weightControls.find('TextInput').props().value).toEqual(
        exercise.sets[1].weight.toString()
      );
      expect(repsControls.find('TextInput').props().value).toEqual(
        exercise.sets[1].reps.toString()
      );
    });

    it('changes input(s) state using the TextInput', () => {
      const wrapper = shallow(
        <EditSetsWithControls
          day={day}
          dispatch={dispatch}
          exerciseKey={exerciseKey}
          exercisesCount={1}
          maxSetId=""
        />
      );
      const weightControls = getInputControls(wrapper, 0);
      const repsControls = getInputControls(wrapper, 1);

      const _checkStateAfterChange = (controls, stateLabel, value) => {
        controls
          .find('TextInput')
          .props()
          .onChangeText(value.toString());

        expect(wrapper.state()[stateLabel]).toEqual(value);

        controls
          .find('TextInput')
          .props()
          .onChangeText('');

        expect(wrapper.state()[stateLabel]).toEqual(-1);
      };

      _checkStateAfterChange(weightControls, 'weight', 50.0);
      _checkStateAfterChange(repsControls, 'reps', 5);
    });

    it.skip('handles empty TextInput', () => {
      // TODO write test
    });

    it('uses -2, -1, +1, +2 buttons for reps', () => {
      const changeRepsValue = (buttons, buttonIndex) => {
        buttons
          .at(buttonIndex)
          .props()
          .onPress();
      };

      const wrapper = shallow(
        <EditSetsWithControls
          day={day}
          dispatch={dispatch}
          exerciseKey={exerciseKey}
          exercisesCount={1}
          maxSetId=""
        />
      );
      const weightControls = getInputControls(wrapper, 1);

      const buttons = weightControls.find('withTheme(TouchableRipple)');

      changeRepsValue(buttons, 0);
      expect(wrapper.state().reps).toEqual(defaultReps - 2);

      changeRepsValue(buttons, 1);
      expect(wrapper.state().reps).toEqual(defaultReps - 2 - 1);

      changeRepsValue(buttons, 2);
      expect(wrapper.state().reps).toEqual(defaultReps - 2 - 1 + 1);

      changeRepsValue(buttons, 3);
      expect(wrapper.state().reps).toEqual(defaultReps - 2 - 1 + 1 + 2);
    });

    it('uses -1.0, -0.5, +0.5, +1.0 buttons for weight', () => {
      const changeWeightValue = (buttons, buttonIndex) => {
        buttons
          .at(buttonIndex)
          .props()
          .onPress();
      };

      const wrapper = shallow(
        <EditSetsWithControls
          day={day}
          dispatch={dispatch}
          exerciseKey={exerciseKey}
          exercisesCount={1}
          maxSetId=""
        />
      );
      const weightControls = getInputControls(wrapper, 0);

      const buttons = weightControls.find('withTheme(TouchableRipple)');

      changeWeightValue(buttons, 0);
      expect(wrapper.state().weight).toEqual(defaultWeight - 0.5);

      changeWeightValue(buttons, 1);
      expect(wrapper.state().weight).toEqual(defaultWeight - 0.5 - 1.0);

      changeWeightValue(buttons, 2);
      expect(wrapper.state().weight).toEqual(defaultWeight - 0.5 - 1.0 + 1.0);

      changeWeightValue(buttons, 3);
      expect(wrapper.state().weight).toEqual(
        defaultWeight - 0.5 - 1.0 + 1.0 + 0.5
      );
    });
  });

  describe('EditSetActionButtons and back button', () => {
    const wrapper = shallow(
      <EditSetsWithControls
        day={day}
        dispatch={dispatch}
        exerciseKey={exerciseKey}
        exercisesCount={1}
        maxSetId=""
      />
    );

    beforeEach(() => {
      jest.clearAllMocks();
      wrapper.setState({ selectedId: '' });
    });

    // eslint-disable-next-line flowtype/no-weak-types
    const getActionButtons = (parent: Object) => {
      const inputControls = parent
        .find('withTheme(EditSetActionButtons)')
        .dive();

      const Component = inputControls.props().children();
      return shallow(React.cloneElement(Component, { theme }));
    };

    it('switches between Add and Update text if a set is selected', () => {
      expect(
        getActionButtons(wrapper)
          .find('withTheme(Button)')
          .at(0)
          .children()
          .text()
      ).toEqual('Add');

      wrapper.setState({ selectedId: exercise.sets[0].id });

      expect(
        getActionButtons(wrapper)
          .find('withTheme(Button)')
          .at(0)
          .children()
          .text()
      ).toEqual('Update');
    });

    it('switches Delete button to enabled/disabled depending on set selection', () => {
      const deleteButton = getActionButtons(wrapper)
        .find('withTheme(Button)')
        .at(1);

      // Nothing selected
      expect(deleteButton.props().disabled).toBe(true);

      wrapper.setState({ selectedId: exercise.sets[0].id });

      // Something selected means delete button is enabled
      expect(deleteButton.props().disabled).toBe(true);
    });

    it('handles back button if an option is selected', () => {
      wrapper.setState({ selectedId: exercise.sets[0].id });

      // Something is selected, handle it
      expect(wrapper.instance().onBackButtonPressAndroid()).toBe(true);
      expect(wrapper.state().selectedId).toEqual('');

      // Default behavior if nothing is selected
      expect(wrapper.instance().onBackButtonPressAndroid()).toBe(false);
    });

    it.skip('adds a set and dismiss the keyboard', () => {
      expect(Keyboard.dismiss).not.toBeCalled();

      wrapper.instance()._onAddSet();

      expect(Keyboard.dismiss).toBeCalled();

      // TODO test rest of logic inside here
    });

    it('deletes a set and dismiss the keyboard', () => {
      expect(Keyboard.dismiss).not.toBeCalled();

      wrapper.instance()._onDeleteSet();

      expect(Keyboard.dismiss).toBeCalled();
      expect(deleteSet).toHaveBeenCalledTimes(1);
    });
  });
});
