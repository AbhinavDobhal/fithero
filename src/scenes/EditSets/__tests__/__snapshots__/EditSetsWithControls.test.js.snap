// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditSetsWithControls Weight units renders kgs or lbs 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -132,1 +132,1 @@
-             Weight (kgs)
+             Weight (lb)
@@ -250,1 +250,1 @@
-                 value="90"
+                 value="198.42"
@@ -945,1 +945,1 @@
-               100
+               220.46
@@ -967,1 +967,1 @@
-                 kgs
+                 lb
@@ -1131,1 +1131,1 @@
-               90
+               198.42
@@ -1153,1 +1153,1 @@
-                 kgs
+                 lb
`;
